File: rs/earthbucks_lib/tests/block_test.rs
use earthbucks_lib::block::Block;
use earthbucks_lib::header::Header;
use earthbucks_lib::iso_buf_reader::IsoBufReader;
use earthbucks_lib::tx::Tx;

#[test]
fn test_to_buffer_writer() {
    let header = Header {
        version: 1,
        prev_block_id: [0; 32],
        merkle_root: [0; 32],
        timestamp: 0,
        block_num: 0,
        target: [0; 32],
        nonce: [0; 32],
        work_ser_algo: 0,
        work_ser_hash: [0; 32],
        work_par_algo: 0,
        work_par_hash: [0; 32],
    };
    let tx = Tx::new(1, vec![], vec![], 1);
    let block = Block::new(header, vec![tx]);
    let bw = block.to_buffer_writer();
    assert!(!bw.to_iso_buf().is_empty());
}

#[test]
fn test_to_iso_buf_and_from_iso_buf() {
    let header = Header {
        version: 1,
        prev_block_id: [0; 32],
        merkle_root: [0; 32],
        timestamp: 0,
        block_num: 0,
        target: [0; 32],
        nonce: [0; 32],
        work_ser_algo: 0,
        work_ser_hash: [0; 32],
        work_par_algo: 0,
        work_par_hash: [0; 32],
    };
    let tx = Tx::new(1, vec![], vec![], 1);
    let block1 = Block::new(header, vec![tx]);
    let buf = block1.to_iso_buf();
    let block2 = Block::from_iso_buf(buf).unwrap();
    assert_eq!(block1.header.version, block2.header.version);
    assert_eq!(block1.txs[0].version, block2.txs[0].version);
}

#[test]
fn test_from_iso_buf_reader() {
    let header = Header {
        version: 1,
        prev_block_id: [0; 32],
        merkle_root: [0; 32],
        timestamp: 0,
        block_num: 0,
        target: [0; 32],
        nonce: [0; 32],
        work_ser_algo: 0,
        work_ser_hash: [0; 32],
        work_par_algo: 0,
        work_par_hash: [0; 32],
    };
    let tx = Tx::new(1, vec![], vec![], 1);
    let block1 = Block::new(header, vec![tx]);
    let buf = block1.to_iso_buf();
    let mut br = IsoBufReader::new(buf);
    let block2 = Block::from_iso_buf_reader(&mut br).unwrap();
    assert_eq!(block1.header.version, block2.header.version);
    assert_eq!(block1.txs[0].version, block2.txs[0].version);
}

#[test]
fn test_is_genesis() {
    let header = Header {
        version: 1,
        prev_block_id: [0; 32],
        merkle_root: [0; 32],
        timestamp: 0,
        block_num: 0,
        target: [0; 32],
        nonce: [0; 32],
        work_ser_algo: 0,
        work_ser_hash: [0; 32],
        work_par_algo: 0,
        work_par_hash: [0; 32],
    };
    let tx = Tx::new(1, vec![], vec![], 0);
    let block = Block::new(header, vec![tx]);
    assert!(block.header.is_genesis());
}

File: ts/npm-earthbucks-ebx-lib/test/block.test.ts
import { describe, expect, test, beforeEach, it } from "vitest";
import { Header } from "../src/header.js";
import { Block } from "../src/block.js";
import { Tx } from "../src/tx.js";
import { BufWriter } from "@webbuf/rw";
import { BufReader } from "@webbuf/rw";
import { U8, U16BE, U32BE, U64BE, U128BE, U256BE } from "@webbuf/numbers";
import { WORK_SER_ALGO_NUM } from "../src/work-ser-algo.js";
import { WORK_PAR_ALGO_NUM } from "../src/work-par-algo.js";

describe("Block", () => {
  test("toBufWriter", () => {
    const bh = new Header();
    const tx = new Tx(new U8(0), [], [], new U32BE(0n));
    const block = new Block(bh, [tx]);
    const bw = block.toBufWriter(new BufWriter());
    expect(bw.toBuf().length).toBeGreaterThan(0);
  });

  test("toBuf", () => {
    const bh = new Header();
    const tx = new Tx(new U8(0), [], [], new U32BE(0n));
    const block = new Block(bh, [tx]);
    const u8vec = block.toBuf();
    expect(u8vec.length).toBeGreaterThan(0);
  });

  test("fromBufReader", () => {
    const bh = new Header();
    const tx = new Tx(new U8(0), [], [], new U32BE(0n));
    const block = new Block(bh, [tx]);
    const bw = block.toBufWriter(new BufWriter());
    const buf = bw.toBuf();
    const br = new BufReader(buf);
    const block2 = Block.fromBufReader(br);
    expect(block2.header.version.n).toBe(bh.version.n);
    expect(block2.header.prevBlockId.buf.toString("hex")).toEqual(
      bh.prevBlockId.buf.toString("hex"),
    );
    expect(block2.header.rootMerkleTreeId).toEqual(bh.rootMerkleTreeId);
    expect(block2.header.timestamp.bn).toBe(bh.timestamp.bn);
    expect(block2.header.target).toEqual(bh.target);
    expect(block2.header.nonce).toEqual(bh.nonce);
    expect(block2.header.blockNum.bn).toBe(bh.blockNum.bn);
  });

  test.skip("isGenesis", () => {
    const bh = new Header({
      workSerAlgo: new U16BE(WORK_SER_ALGO_NUM.blake3_3),
      workParAlgo: new U16BE(WORK_PAR_ALGO_NUM.algo1627),
    });
    const tx = new Tx(new U8(0), [], [], new U32BE(0n));
    const block = new Block(bh, [tx]);
    expect(block.isGenesis()).toBe(true);
  });
});
